src/main/java/minijava/lexer/SimpleLexer.java

```diff
@@ -0,0 +1,410 @@
+    }
+  }
+
+  private Token parseStar() {
+    switch (input.current()) {
+      case '=':
+        input.next();
+        return createToken(Terminal.RESERVED_OPERATORS, "*=");
+      default:
+        input.next();
+        return createToken(Terminal.MULTIPLY, "*");
+    }
+  }
+
+  private Token parsePipe() {
+    switch (input.current()) {
+      case '|':
+        input.next();
+        return createToken(Terminal.OR, "||");
+      default:
+        return createToken(Terminal.RESERVED_OPERATORS, "|");
+    }
+  }
+
+  private Token parseIdent() {
+    StringBuilder builder = new StringBuilder();
+    builder.appendCodePoint(input.current());
+    char cur = (char) (int) input.next();
+    while (input.current() != -1 && (isAlphabet(cur) || cur == '_' || Character.isDigit(cur))) {
+      builder.appendCodePoint(cur);
+      cur = (char) (int) input.next();
+    }
+    return createToken(Terminal.IDENT, builder.toString());
+  }
+
+  private boolean isAlphabet(char c) {
+    return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
+  }
+
+  private Token createToken(Terminal terminal, String content) {
+    int stringId = stringTable.getStringId(content);
+    if (terminal != Terminal.IDENT) {
+      return new Token(terminal, location, stringId, stringTable);
+    }
+    Terminal actualTerminal = Terminal.IDENT;
+    if (keywordTerms.containsKey(stringId)) {
+      actualTerminal = keywordTerms.get(stringId);
+    }
+    return new Token(actualTerminal, location, stringId, stringTable);
+  }
+
+  private Token nextToken() {
+    if (current != null && current.isEOF()) {
+      return current;
+    }
+    if (lookAheadBuffer.size() > 0) {
+      current = lookAheadBuffer.get(0);
+      lookAheadBuffer.remove(0);
+    } else {
+      current = parseNextToken();
+    }
+    return current;
+  }
+
+  @Override
+  public Token current() {
+    if (current == null) {
+      current = parseNextToken();
+      if (current.isEOF()) {
+        numberOfEOFs++;
+      }
+    }
+    return current;
+  }
+
+  @Override
+  public Token lookAhead(int lookAhead) {
+    if (lookAhead <= 0) {
+      return current();
+    }
+    if (current().isEOF()) {
+      return current();
+    }
+    Token curToken = current();
+    while (lookAheadBuffer.size() < lookAhead && !curToken.isEOF()) {
+      curToken = parseNextToken();
+      lookAheadBuffer.add(curToken);
+    }
+    if (lookAheadBuffer.size() >= lookAhead) {
+      return lookAheadBuffer.get(lookAhead - 1);
+    } else {
+      return lookAheadBuffer.get(lookAheadBuffer.size() - 1);
+    }
+  }
+
+  @Override
+  public StringTable getStringTable() {
+    return stringTable;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return numberOfEOFs < 1;
+  }
+
+  @Override
+  public Token next() {
+    current = nextToken();
+    if (!lookAheadBuffer.isEmpty()) {
+      lookAheadBuffer.remove(0);
+    }
+    if (current.isEOF()) {
+      numberOfEOFs++;
+    }
+    return current;
+  }
+}
```

Comment:
Some points:

1. As above, we should probably split off lookahead logic into a separate `Iterable`...
2. nextToken(), next(), lookAhead() and current() are really conflated and I'm not sure it's bug-free, but let's see it with tests
3. Also wasn't this supposed to be a Table-driven lexer?

Otherwise I'm fine with this.